src/main/java/com/zaxxer/hikari/util/ConcurrentBag.java

```diff
@@ -301,7 +307,9 @@ public boolean reserve(final T bagEntry)
    public void unreserve(final T bagEntry)
    {
       if (bagEntry.compareAndSet(STATE_RESERVED, STATE_NOT_IN_USE)) {
-         synchronizer.signal();
+         if (bagEntry.getState() == STATE_NOT_IN_USE) {
```

Comment:
unreserve() method rare case in real lifecycle of connection. but requite() and add() should recheck state after free and it will improve performance.